walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -11,134 +11,200 @@
    */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
     int duration = (int) request.getDuration();
     Collection<String> attendees = request.getAttendees();
     Collection<String> optionalAttendees = request.getOptionalAttendees();
 
     Collection<TimeRange> availableTimes = new ArrayList<>();
-    Collection<TimeRange> availableTimesOptionals = new ArrayList<>();
+    List<Integer> availableTimesOptionals = new ArrayList<>();
 
     availableTimes = getAvailableTimes(attendees, events, duration);
+    availableTimesOptionals = getAvailableTimesForOptionals(optionalAttendees, events, duration);
 
-    // If there is no optional attendees only consider required attendees
-    if (optionalAttendees.isEmpty()) {
-      return availableTimes;
-    }
+    Collection<TimeRange> overlaps =
+        findOverlaps(availableTimes, availableTimesOptionals, duration, optionalAttendees.size());
 
-    // Calculate best times for optional attendees
-    availableTimesOptionals = getAvailableTimes(optionalAttendees, events, duration);
+    return overlaps;
+  }
 
-    // If there is no required attendees, return time which works for optionals
-    if (attendees.isEmpty()) {
-      return availableTimesOptionals;
-    }
+  /**
+   * Finds avaliable times considering the required atteendee list, existing events and event
+   * duration
+   */
+  private Collection<TimeRange> getAvailableTimes(
+      Collection<String> attendees, Collection<Event> events, int duration) {
+    List<TimeRange> busyTimes = new ArrayList<>();
+    Collection<TimeRange> freeTimes;
 
-    // Find times that works both for required and optional attendees
-    Collection<TimeRange> availableTimesBothAttendees =
-        intersection(availableTimesOptionals, availableTimes, duration);
+    // finds time ranges that does not work for at least one required attendee
+    for (Event event : events) {
+      TimeRange eventTimeRange = event.getWhen();
 
-    // Return time if any which works for both required and optional attendees
-    if (!availableTimesBothAttendees.isEmpty()) {
-      return availableTimesBothAttendees;
+      Collection<String> eventAttendees = event.getAttendees();
+      if (!Collections.disjoint(attendees, eventAttendees)) {
+        busyTimes.add(eventTimeRange);
+      }
     }
 
-    return availableTimes;
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+    freeTimes = findFreeTimesFromBusy(busyTimes, duration);
+
+    return freeTimes;
   }
 
-  /** Finds avaliable times for the given atteendee list, input collection and event duration */
-  private Collection<TimeRange> getAvailableTimes(
-      Collection<String> attendees, Collection<Event> events, int duration) {
-    // TODO: Make this method more optimal
-    Collection<TimeRange> availableTimes = new ArrayList<>();
+  /** Calculates for each minute in a day how many optional attendee are free */
```

Comment:
So this should return a list `j` of size 1440, with `j[i]` meaning the number of optional attendees are free at minute `i`? The function comment is a little unclear.